packages/contracts/hardhat/tasks/check-balance-for-deploy.ts

```diff
@@ -0,0 +1,122 @@
+import fs from 'node:fs'
+import path from 'node:path'
+import { TASK_DEPLOY } from 'hardhat-deploy'
+import { task } from 'hardhat/config'
+import type { HardhatRuntimeEnvironment } from 'hardhat/types'
+import type { DeployFunction } from 'hardhat-deploy/types'
+
+export const checkBalanceForDeploy = task(TASK_DEPLOY, async (_args, hre: HardhatRuntimeEnvironment, runSuper) => {
+  if (hre.network.name === 'hardhat') {
+    return runSuper(_args)
+  }
+  const gasUsedTotal = await getEstimatedDeploymentGas(hre)
+  const gasPrice = await hre.ethers.provider.getGasPrice()
+  const estimatedDeploymentPrice = gasPrice.mul(gasUsedTotal).mul(120).div(100)
+  const availableBalance = await getDeployerBalance(hre)
+
+  const requiredBalance = hre.ethers.utils.formatEther(estimatedDeploymentPrice)
+  const currentBalance = hre.ethers.utils.formatEther(availableBalance)
+  console.log(`Deployer\'s account balance: ${currentBalance}, required: ${requiredBalance}.`)
+
+  const hasEnoughBalanceToDeploy = availableBalance.gte(estimatedDeploymentPrice)
+  if (!hasEnoughBalanceToDeploy) {
+    throw new Error('There are insufficient funds in the deployer\'s account.')
+  }
+  return runSuper(_args)
+})
+
+async function getDeployerBalance(hre: HardhatRuntimeEnvironment) {
+  const accounts = await hre.ethers.getSigners()
+  if (accounts.length === 0) {
+    throw new Error('No accounts data found. Check your env. vars!')
+  }
+  const balance = await accounts[0].getBalance()
+  return balance
+}
+
+async function getEstimatedDeploymentGas(hre: HardhatRuntimeEnvironment) {
+  const deployFunctions = await getDeployFunctions(hre)
+  let gasUsedTotal = hre.ethers.BigNumber.from(0)
+  for (const deployFunction of deployFunctions) {
+    const gasUsed = await getGasUsedForDeployment(deployFunction, hre)
+    gasUsedTotal = gasUsedTotal.add(gasUsed)
+  }
+  return gasUsedTotal
+}
+
+async function getGasUsedForDeployment(deployFunction: DeployFunction, hre: HardhatRuntimeEnvironment) {
+  const [contractName] = deployFunction.tags ?? []
+  if (!contractName) {
+    throw new Error('Wrong deployment, no tags?')
+  }
+
+  const deploymentName = await getDeploymentName(contractName, hre)
+  if (!deploymentName) {
+    return await getEstimatedGasPrice(hre)
+  }
```

Comment:
The `getGasUsedForDeployment` return gas used for deployment, but `getEstimatedGasPrice` returns gas for all deployments. Does it correct?

Reply:
getEstimatedGasPrice returns gas for the single deployment, for contract which is never been deployed yet (it's estimates the price using the median value). I will rename it and add the comments